modules/002-deckhouse/webhooks/validating/d8-cluster-configuration

```diff
@@ -66,56 +53,69 @@ EOF
-    endpointsCount="$(context::jq -r '.snapshots.endpoints[].filterResult.count')"
-    oldDefaultCRI="$(context::jq -r '.review.request.oldObject.data."cluster-configuration.yaml"' | base64 -d | yq eval '.defaultCRI // "Containerd"' -)"
-    newDefaultCRI="$(context::jq -r '.review.request.object.data."cluster-configuration.yaml"' | base64 -d | yq eval '.defaultCRI // "Containerd"' -)"
+    oldConfig=$(context::jq -r '.review.request.oldObject.data."cluster-configuration.yaml"' | base64 -d)
+    newConfig=$(context::jq -r '.review.request.object.data."cluster-configuration.yaml"' | base64 -d)
+    unsafeMode=$(context::jq -r '.review.request.object.metadata.annotations."deckhouse.io/allow-unsafe"')
 
-    if [[ ("${newDefaultCRI}" != "${oldDefaultCRI}") && ("${endpointsCount}" -lt 3) ]]; then
-      cat <<EOF > "$VALIDATING_RESPONSE_PATH"
-{"allowed":true, "warnings":["it is disruptive to change defaultCRI type for cluster with apiserver endpoints < 3"]}
-EOF
-      return 0
-    fi
-  fi
+    if [ "${unsafeMode}" == null ]; then
+      # podSubnetNodeCIDRPrefix cannot be changed in a running cluster in a safe mode
+      oldPrefix=$(yq eval '.podSubnetNodeCIDRPrefix' - <<< "$oldConfig")
+      newPrefix=$(yq eval '.podSubnetNodeCIDRPrefix' - <<< "$newConfig")
+      if [ "${oldPrefix}" != "${newPrefix}" ]; then
+        echo '{"allowed":false, "message":"it is forbidden to change podSubnetNodeCIDRPrefix in a running cluster"}' > "$VALIDATING_RESPONSE_PATH"
+        return 0
+      fi
 
-  # check CRI compatibility with k8s 1.24, docker CRI was deprecated there
-  if context::jq -e -r '.review.request.operation == "UPDATE"' >/dev/null 2>&1; then
-    k8sVersion="$(context::jq -r '.review.request.object.data."cluster-configuration.yaml"' | base64 -d | yq eval '.kubernetesVersion // "1.21"' -)"
+      # podSubnetNodeCIDRPrefix cannot be changed in a running cluster in a safe mode
+      oldPrefix=$(yq eval '.podSubnetNodeCIDRPrefix' - <<< "$oldConfig")
+      newPrefix=$(yq eval '.podSubnetNodeCIDRPrefix' - <<< "$newConfig")
+      if [ "${oldPrefix}" != "${newPrefix}" ]; then
+        echo '{"allowed":false, "message":"it is forbidden to change podSubnetNodeCIDRPrefix in a running cluster"}' > "$VALIDATING_RESPONSE_PATH"
+        return 0
+      fi
 
-    if [ "$k8sVersion" == "1.24" ]; then
-      defaultCRI="$(context::jq -r '.review.request.object.data."cluster-configuration.yaml"' | base64 -d | yq eval '.defaultCRI // "Containerd"' -)"
+      # podSubnetNodeCIDRPrefix cannot be changed in a running cluster in a safe mode
+      oldPrefix=$(yq eval '.podSubnetNodeCIDRPrefix' - <<< "$oldConfig")
+      newPrefix=$(yq eval '.podSubnetNodeCIDRPrefix' - <<< "$newConfig")
+      if [ "${oldPrefix}" != "${newPrefix}" ]; then
+        echo '{"allowed":false, "message":"it is forbidden to change podSubnetNodeCIDRPrefix in a running cluster"}' > "$VALIDATING_RESPONSE_PATH"
+        return 0
+      fi
```

Comment:
you eval oldPrefix and newPrefix 3 times. I think this is a copy/paste error

Reply:
Fixed, thanks.